### PR TITLE
`Development`: Add documentation for build job cancellation

### PR DESCRIPTION
**Pull request by mateusmm01**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

### Motivation and Context

The System Design for the  Build Job Cancellation is a complex process and is missing in the documentation.

### Description

First I described the Hazelcast Data Structures. Then I added and described two diagrams (Class and Communication Diagram)

### Steps for Testing

https://artemis-platform--8410.org.readthedocs.build/en/8410/dev/system-design.html#hazelcast-distributed-data-structures
https://artemis-platform--8410.org.readthedocs.build/en/8410/dev/system-design.html#build-job-cancellation

#### Code Review
- [x] Code Review 1
- [ ] Code Review 2

## Summary by CodeRabbit

- **New Features**
	- Introduced Hazelcast distributed data structures for efficient build job processing in the CI system.
	- Enhanced user interaction capabilities through REST endpoints and WebSocket connections.

### CODE REVIEW

File: docs/dev/system-design/integrated-code-lifecycle/integrated-code-lifecycle-system-design.rst [hunk 1]
```diff
@@ -142,3 +142,67 @@ The ephemeral nature of Docker containers allows the ``BuildJobExecutionService`
...
+   :align: center
+   :width: 800
+   :alt: Build Job Cancellation
+
+
+   Build Job Cancellation Class Diagram
+
+The build job cancellation mechanism is implemented using Hazelcast's distributed data structures.
+As shown in the `Build Job Cancellation Class Diagram` we have services such as ``SharedQueueManagementService`` that run on the core Artemis node and are responsible for managing Hazelcast's distributed datatypes.
+The real processing of each build job however runs independently on each build agent, therefore we have separate services such as ``SharedQueueProcessingService`` in a build agent package .
```
Comment (rstief) [inline docs/dev/system-design/integrated-code-lifecycle/integrated-code-lifecycle-system-design.rst:177]:
```suggestion
As shown in the `Build Job Cancellation Class Diagram` we have services such as the ``SharedQueueManagementService`` that run on the core Artemis node and are responsible for managing Hazelcast's distributed datatypes.
However, the real processing of each build job runs independently on each build agent. Therefore we have separate services such as the ``SharedQueueProcessingService`` in a build agent package .
```

File: docs/dev/system-design/integrated-code-lifecycle/integrated-code-lifecycle-system-design.rst [hunk 2]
```diff
@@ -142,3 +142,67 @@ The ephemeral nature of Docker containers allows the ``BuildJobExecutionService`
...
+
+The processing of build jobs gets triggered by the insertion or removal of jobs from the ``buildJobQueue``.
+The queue listener in the ``SharedQueueProcessingService`` is then responsible for invoking the ``checkAvailabilityAndProcessNextBuild()`` method.
+If a build agent has available threads, the job is moved from the ``buildJobQueue`` to the ``processingJobs``.
+Subsequently, the ``processBuild(LocalCIBuildJobQueueItem buildJob)`` method is called, which in turn invokes ``executeBuildJob(LocalCIBuildJobQueueItem buildJob)`` in the ``BuildJobManagementService``.
+This method is tasked with asynchronously executing this build job in the ``BuildJobExecutionService``, by creating a future.
```
Comment (rstief) [inline docs/dev/system-design/integrated-code-lifecycle/integrated-code-lifecycle-system-design.rst:183]:
```suggestion
This method is tasked with asynchronously executing the build job in the ``BuildJobExecutionService``, by creating a future.
```

File: docs/dev/system-design/integrated-code-lifecycle/integrated-code-lifecycle-system-design.rst [hunk 3]
```diff
@@ -142,3 +142,67 @@ The ephemeral nature of Docker containers allows the ``BuildJobExecutionService`
...
+Upon creating the new future, it is added to a local ``Map<String, Future<LocalCIBuildResult>> runningFutures``,
+unique to each build agent and containing only the futures of that agent.
```
Comment (rstief) [inline docs/dev/system-design/integrated-code-lifecycle/integrated-code-lifecycle-system-design.rst:185]:
```suggestion
Upon creating the new future, it is added to a local ``Map<String, Future<LocalCIBuildResult>> runningFutures``, which is unique to each build agent and contains only the futures of that agent.
```

File: docs/dev/system-design/integrated-code-lifecycle/integrated-code-lifecycle-system-design.rst [hunk 4]
```diff
@@ -142,3 +142,67 @@ The ephemeral nature of Docker containers allows the ``BuildJobExecutionService`
...
+   :alt: Build Job Cancellation
+
+   Build Job Cancellation Communication Diagram for a Processing Build Job
+
+
+The above Communication Diagram sets the stage for understanding the cancellation mechanism.
+Upon receiving a cancellation request through the ``@DeleteMapping("cancel-job/{buildJobId}")`` endpoint in either the ``AdminBuildJobQueueResource`` or ``BuildJobQueueResource``,
+the ``cancelBuildJob(String buildJobId)`` method within the ``SharedQueueManagementService`` is activated.
+Initially, it checks whether the build job is queued but not yet in process. If so, it is simply removed from the ``buildJobQueue``.
+The process becomes more complex for jobs already in process, involving the ``canceledBuildJobsTopic`` ITopic.
```
Comment (rstief) [inline docs/dev/system-design/integrated-code-lifecycle/integrated-code-lifecycle-system-design.rst:201]:
```suggestion
The process becomes more complex for jobs already in process, involving the ``canceledBuildJobsTopic``  ``ITopic``.
```

File: docs/dev/system-design/integrated-code-lifecycle/integrated-code-lifecycle-system-design.rst [hunk 5]
```diff
@@ -142,3 +142,67 @@ The ephemeral nature of Docker containers allows the ``BuildJobExecutionService`
...
+A cancellation message for the build job ID is broadcast to all build agents, with the ``BuildJobManagementService`` listening for these messages via Hazelcast's ``MessageListener``.
+Every build agent then checks its ``runningFutures`` map to see if it holds the specific build job future.
+Should a build agent currently process the specified job, it executes ``cancelBuildJob(String buildJobId)`` within the ``BuildJobManagementService``.
+This step involves canceling the future and recording the job ID in ``Set<String> cancelledBuildJobs``.
+The cancellation triggers an Exception which is caught and managed in ``processBuild(...)`` within ``SharedQueueProcessingService``.
```
Comment (rstief) [inline docs/dev/system-design/integrated-code-lifecycle/integrated-code-lifecycle-system-design.rst:206]:
```suggestion
The cancellation triggers an Exception which is caught and managed in ``processBuild(...)`` within the ``SharedQueueProcessingService``.
```

**Review by rstief — COMMENTED** [review]
LGTM, some small comments on grammar :)

**Review by laurenzfb — APPROVED** [review]
lgtm

**Pull request merged**